pkg/controller/metrics/metrics.go

```diff
@@ -225,6 +235,25 @@ func (mc *Calculator) Start(ctx context.Context) error {
 						mcLog.Infof("cleared metric: %v", metricClusterDeploymentSyncsetPaused)
 					}
 				}
+
+				// Check if cluster is transitioning
+				hibernatingCond := controllerutils.FindClusterDeploymentCondition(cd.Status.Conditions, hivev1.ClusterHibernatingCondition)
+				readyCond := controllerutils.FindClusterDeploymentCondition(cd.Status.Conditions, hivev1.ClusterReadyCondition)
+				poolRef := ""
+				if cd.Spec.ClusterPoolRef != nil {
+					poolRef = cd.Spec.ClusterPoolRef.PoolName
+				}
+				if hibernatingCond.Reason == hivev1.ResumingOrRunningHibernationReason &&
```

Comment:
This seems an odd way/place to be observing/calculating this metric. If I understand the logic correctly, we're only reporting *during* the power state transition; and we're reporting the time taken *so far*; and we're doing that *every time this loop runs*. So for example if it takes 8m to resume and this loop runs every 2m and the first time it kicks in is 1m into my resume, we would end up reporting *four* intervals (1m, 3m, 5m, 7m) when what we actually wanted was only one (8m).

It would seem to make more sense to report this metric in the hibernation controller, at the exact point where we change the hibernating/ready reason to Hibernating/Running, respectively. (The `time.Since` should be based on the *other* condition's `LastTransitionTime`, as you've done here, but being careful to calculate that *before* the status update.) That way we only report one metric per transition, and it represents the entire time frame.

Reply:
The intent of this change was for admins to be aware of clusters that take more time to transition than trends suggest (so for situations where clusters are stuck hibernating or resuming) - so I can't visualize how it would work in hibernation controller. Therefore I kept the metric in metrics controller - where we loop over all cds and log similar metrics, but I am open to suggestions.

Reply:
Discussed in scrum today. Summary:
- Keep the histogram(s -- one for hibernating, one for resuming) but move their computation to the hibernation controller as suggested, only tracking *completed* state transitions.
- Add two new pairs of metrics -- one for hibernating, one for resuming:
  - A gauge that counts the number of CDs *currently* transitioning
  - A gauge that sums the time spent so far in that transition for all such CDs

  This one needs to be computed here in the async loop. The use case is to divide these two numbers (yielding the "average time in active power state transitions") and get worried if that average starts to increase, indicating that one or more transitions are taking longer than they should.